WIP-0009: submit draft (#40)
witnet/WIPs: wip-0009.md

<pre>
    WIP: WIP-0009
    Layer: Consensus (hard fork)
    Title: Adjust mining probability
    Authors: Mario Cao <mario@witnet.foundation>
             Gorka Irazoqui <gorka.irazoki@gmail.com>
    Discussions-To: `#dev-general` channel on Witnet Community's Discord server
    Status: Draft
    Type: Standards Track
    Created: 2021-01-26
    License: BSD-2-Clause
</pre>

## Abstract

> This proposal adjusts the mining algorithm to limit the probability of non-reputed and non-active nodes mining a block.

Reviewer: ```suggestion
This proposal adjusts the mining algorithm to limit the probability of non-reputed and non-active identities for mining a block.
```

The goal of this proposal is to mitigate potential network abuses that could be carried by malicious users controlling a large number of nodes. This could pose a serious threat to the network availability and security.

## Motivation and Rationale

after more than 3 months of the Witnet mainnet launch, the network has faced episodes of a bigger transaction latency than expected. In periods with a mempool full of transactions, blocks not using the complete space appeared in the Witnet network. This reduces the efficiency of the system, thus increasing indirectly the transaction latency perceived by users. Under a more critical scenario, this problem could pose an availability threat in terms of a [denial-of-service attack][DoS], in which a large number of malicious nodes could be mining empty blocks or, even, endanger the network stability by affecting internal network metrics (e.g. Active Reputation Set).

> A thorough investigation concluded that the majority of these blocks (if not all) were produced by non-reputed and non-active nodes (i.e. newcomers to the network). Regardless of the intentionality of this behaviour, these blocks were appearing in a much higher volume than the newcomer mining probability that was designed in the protocol.

Reviewer: ```suggestion
A thorough investigation concluded that most, if not all, these blocks were produced by non-reputed and non-active nodes (i.e. newcomers to the network). Regardless of whether this is an intentional or accidental behavior, these blocks were appearing in a much higher number than it is to be expected from the newcomer mining probability as provided by the protocol.
```

The current mining algorithm can be summarized as follows:

> 1. Miners build a candidate block.

Reviewer: Seems like _candidate_ doesn't exist as an adjective, so I'd rather say _block candidate_ as in _PhD candidate_ meaning _candidate for block consolidation_.

I think this is easier to understand if we try to make the description embody the behavior of a single miner: 
```suggestion
1. Each miner builds a block candidate.
```

Author: 😔

> 2. Miners compute their eligibility by using the [VRF]-based [RandPoE] algorithm with by using the ARS census as the difficulty parameter (labelled as N).

Reviewer: ```suggestion
2. Each miner computes its own eligibility by using the [VRF]-based [RandPoE] algorithm by using the ARS size as the difficulty parameter (labelled as `N`).
```

> 3. If eligible (and no better candidate was received), miners will broadcast their candidate blocks.

Reviewer: ```suggestion
3. May a miner be eligible, and had it not received a better block candidate, it will broadcast its own candidate.
```

> Upon reception of multiple candidates, the winner will be selected as follows:

Reviewer: ```suggestion
Upon reception of multiple candidates, the winner is selected as follows:
```

> 1. Candidate blocks from reputed nodes are prioritized over those from non-reputed nodes.

Reviewer: ```suggestion
1. Block candidates signed by reputed identities are prioritized over those from non-reputed identities.
```

> 2. If there are still multiple candidates, the one with the lowest VRF is chosen.

Reviewer: For the sake of completeness, is it worth mentioning that in case of a tie, block hashes are preferred in lexicographical order?

Author: The VRF output is a number... 
I suggest: "2. If there are still multiple candidates, the one with the lowest VRF value is chosen."

Reviewer: Yeah, but I mean that this description doesn't really match what I see in the code (it's missing "slots" and block hash comparison).

Author: There were left out on purpuse for the sake of simplicity. Nevertheless, we added 2 more steps. :)

> The main problem with the aforementioned algorithm is that the eligibility is computed using the ARS as census, but the prioritization is only taking into account the reputed nodes. This led to the increase of the newcomer mining probability. For example, the 18th of January the newcomer mining probability could be computed with the following network conditions:

Reviewer: ```suggestion
The main problem with the described algorithm is that the eligibility is computed using the ARS as a census, but the prioritization is only taking into account the reputed nodes. This misalignment leads to the increase of the newcomer mining probability. For example, on the 18th of January the newcomer mining probability would be close to 37%, as computed from the following network conditions:
```

```
TRS(Total Reputation Set)  = 2000 nodes
ARS(Active Reputation Set) = 6000 nodes
RF(Replication Factor) = 3

> P(non-reputed node mines a block) = P(none reputed node mines a block) = P(single reputed does not mine a block)^TRS = (1- P(single reputed mine a block))^TRS = (1 - RF/ARS)^TRS ~= 37%

Reviewer: ```suggestion
P(non-reputed identity mines a block) = P(no reputed identity mines a block) = P(single reputed identity does not mine a block)^TRS = (1- P(single reputed identity mines a block))^TRS = (1 - RF/ARS)^TRS ~= 37%
```

```

> In the initial mining algorithm design [rand-poe], DoS attacks were carefully taken into account. An upper bound for the probability of an attacker censoring blocks was set to approximately 2%. However, before the mainnet launch some mining parameters were changed to achieve a more fair block reward distribution. It was assummed that the reputation expiration would be slower compared to the activity period (related to the ARS). This assumption did not hold in mainnet as seen in the former example. This mistake led to a significant increase in the newcomer mining probability.

Reviewer: This is the first time mentioning block censorship. I understand that this is linked to the attack in which miners may intentionally mine empty blocks, or exclude transactions from blocks at any degree. Let's try to find a way to connect this paragraph to the former description more explicitly. For example, when you first introduce the attack, you can give it a name and then refer it here.

> In the initial mining algorithm design [rand-poe], DoS attacks were carefully taken into account. An upper bound for the probability of an attacker censoring blocks was set to approximately 2%. However, before the mainnet launch some mining parameters were changed to achieve a more fair block reward distribution. It was assummed that the reputation expiration would be slower compared to the activity period (related to the ARS). This assumption did not hold in mainnet as seen in the former example. This mistake led to a significant increase in the newcomer mining probability.

Reviewer: ```suggestion
In the initial mining algorithm design [rand-poe], DoS attacks were carefully taken into account. An upper bound for the probability of an attacker censoring blocks was set to approximately 2%. However, at a certain point before the mainnet launch some mining parameters were changed to achieve a fairer distribution of block rewards. This change assumed that the reputation expiration would be slower compared to the reputation activity period—which has a big influence in considering whether an identity belongs in the ARS. This assumption did not hold in mainnet as seen in the example above. This mistake led to a significant increase in the newcomer mining probability.
```

Reviewer: For future reference, let's try to link the PR that made RandPoE deviate from its original design.

Summing up, there is a need for adjusting the newcomer mining probability to mitigate the impact of potential DoS attacks that may endanger the availability and the security of the Witnet network.

## Proposal / Specification

> This document proposes to use the same census for the eligibility and the prioritization as initially designed. The previous candidate selection will be modified as follows:

Reviewer: ```suggestion
Following entry into force of this proposal, the same census must be used for the computation of eligibility and the prioritization of block candidates, as initially designed. The new block candidate selection algorithm will be as follows:
```

Upon reception of multiple candidates:

> 1. Candidate blocks from reputed nodes are prioritized over those from non-reputed nodes.

Reviewer: ```suggestion
1. Block candidates from reputed identities are prioritized over those from non-reputed identities.
```

> 2. Candidate blocks from active nodes are prioritized over those from non-active nodes.

Reviewer: ```suggestion
2. Block candidates from active identities are prioritized over those from non-active identities.
```

3. If there are still multiple candidates, the one with the lowest VRF is chosen.

> With the previous changes, the newcomer mining probability is bounded as initially designed in [RandPoe] while the fair distribution over those nodes that have proven some contribution to the network service and security still applies.

Reviewer: ```suggestion
With the previous changes, the newcomer mining probability is bounded as initially designed in [RandPoe] while the fair distribution among those identities that have proven some contribution to the network's service and security still applies.
```

3. May a miner be eligible, and had it not received a better block candidate, it will broadcast its own candidate.

Upon reception of multiple candidates, the winner is selected as follows:
1. Block candidates signed by reputed identities are prioritized over those from non-reputed identities.
2. If there are still multiple candidates, the one with the lowest VRF value is chosen.

The main problem with the described algorithm is that the eligibility is computed using the ARS as a census, but the prioritization is only taking into account the reputed nodes. This misalignment leads to the increase of the newcomer mining probability. For example, on the 18th of January the newcomer mining probability would be close to 37%, as computed from the following network conditions:

```
TRS(Total Reputation Set)  = 2000 nodes
ARS(Active Reputation Set) = 6000 nodes
RF(Replication Factor) = 3

P(non-reputed identity mines a block) = P(no reputed identity mines a block) = P(single reputed identity does not mine a block)^TRS = (1- P(single reputed identity mines a block))^TRS = (1 - RF/ARS)^TRS ~= 37%
```

> In the initial mining algorithm design [rand-poe], DoS attacks were carefully taken into account. An upper bound for the probability of an attacker censoring blocks was set to approximately 2%. However, at a certain point before the mainnet launch some mining parameters were changed to achieve a fairer distribution of block rewards [witnet/witnet-rust#1442]. This change assumed that the reputation expiration would be slower compared to the reputation activity period—which has a big influence in considering whether an identity belongs in the ARS. This assumption did not hold in mainnet as seen in the example above. This mistake led to a significant increase in the newcomer mining probability.

Reviewer: ```suggestion
In the [initial mining algorithm design][rand-poe], DoS attacks were carefully taken into account. An upper bound for the probability of an attacker censoring blocks was set to approximately 2%. However, at a certain point before the mainnet launch some [mining parameters were changed to achieve a fairer distribution of block rewards][witnet/witnet-rust#1442]. This change assumed that the reputation expiration would be slower compared to the reputation activity period—which has a big influence in considering whether an identity belongs in the ARS. This assumption did not hold in mainnet as seen in the example above. This mistake led to a significant increase in the newcomer mining probability.
```

P(non-reputed identity mines a block) = P(no reputed identity mines a block) = P(single reputed identity does not mine a block)^TRS = (1- P(single reputed identity mines a block))^TRS = (1 - RF/ARS)^TRS ~= 37%
```

> In the [initial mining algorithm design][rand-poe], DoS attacks were carefully taken into account. An upper bound for the probability of an attacker censoring blocks was set to approximately 2%. However, at a certain point before the mainnet launch some [mining parameters were changed to achieve a fairer distribution of block rewards][witnet/witnet-rust#1442]. This change assumed that the reputation expiration would be slower compared to the reputation activity period—which has a big influence in considering whether an identity belongs in the ARS. This assumption did not hold in mainnet as seen in the example above. This mistake led to a significant increase in the newcomer mining probability.

Reviewer: `rand-poe` reference won't match `RandPoE` below.
